shared/src/store/action-reducers/utils/calculate-treatments.ts

```diff
@@ -1,255 +1,330 @@
+        // Make all personnel stop treating this patient
+        for (const personnelId of Object.keys(patient.assignedPersonnelIds)) {
+            const personnel = getElement(state, 'personnel', personnelId);
+            delete personnel.assignedPatientIds[patient.id];
+        }
+        patient.assignedPersonnelIds = {};
+        patient.isBeingTreated = false;
+        // Make all material stop treating this patient
+        for (const materialId of Object.keys(patient.assignedMaterialIds)) {
+            const material = getElement(state, 'materials', materialId);
+            delete material.assignedPatientIds[patient.id];
+        }
+        patient.assignedMaterialIds = {};
+    } else if (isPersonnel(element)) {
+        const personnel = getElement(state, 'personnel', element.id);
+        // This personnel doesn't treat any patients anymore
+        for (const patientId of Object.keys(personnel.assignedPatientIds)) {
+            const patient = getElement(state, 'patients', patientId);
+            delete patient.assignedPersonnelIds[personnel.id];
+            patient.isBeingTreated = !isEmpty(patient.assignedPersonnelIds);
+        }
         personnel.assignedPatientIds = {};
-    });
-    materials.forEach((material) => {
+    } else if (isMaterial(element)) {
+        const material = getElement(state, 'materials', element.id);
+        // This material doesn't treat any patients anymore
+        for (const patientId of Object.keys(material.assignedPatientIds)) {
+            const patient = getElement(state, 'patients', patientId);
+            delete patient.assignedMaterialIds[material.id];
+        }
         material.assignedPatientIds = {};
-    });
-    const patients = Object.values(state.patients).filter((patient) => {
-        const visibleStatus = Patient.getVisibleStatus(
-            patient,
-            pretriageEnabled,
-            bluePatientsEnabled
-        );
-        return (
-            patient.position !== undefined &&
-            visibleStatus !== 'black' &&
-            visibleStatus !== 'blue'
-        );
-    });
-    patients.forEach((patient) => {
-        patient.isBeingTreated = false;
-    });
-    if (patients.length === 0) {
-        // Don't treat anyone. No one (alive) is there.
+    }
+}
+
+/**
+ * @param element.position is important:
+ *     - if an element was moved: {@link element.position} needs the new position
+ *     - if treatment of an element should be removed: set {@link element.position} to `undefined`
+ * Also sets {@link Patient.visibleStatusChangedSinceTreatment} back to `false` (if element is a patient)
+ */
+export function updateTreatments(
+    state: Mutable<ExerciseState>,
+    element: Mutable<Material | Patient | Personnel>
+) {
+    // The requirement of this function is neither to result in a "perfect" treatment pattern
+    // nor to be independent from its history (calculating the treatments from scratch is allowed to differ
```

Comment:
The comment is correct, but right now (at least what I implemented).
But recalculating everything from scratch should right now result in the same assignments as with the incremental calculation (as every personnel/material) does not take into account other personnel or material and how much a patient is already treated.

Only (which right now is not the case) if there is a dependency created between personnel/material and wether a patient has already enough treatment, a dependence is created.

Reply:
> The comment is correct.

So this can be resolved?

Reply:
> > The comment is correct.
> 
> So this can be resolved?

You mean in GitHub clicking "resolved"?
The comment says something that is technically correct, but not needed.

Right now you can recalculate everything from scratch and it should result in the exact same assignements as with this incremental calculation.

Reply:
> > > The comment is correct.
> > 
> > 
> > So this can be resolved?
> 
> You mean in GitHub clicking "resolved"?
Yes.

> Right now...

Just because it is like this now doesn't mean that it will stay like this in the future.
This comment's goal was to clarify that it isn't necessary to watch out for this when refactoring or trying to understand this algorithm.

It also depends a bit on whether one only looks at the semantic level or the strict representation of the end states. Our current JSON-Set implementations (e.g. `assignedMaterialIds`) are objects (and therefore have a key order).  Changing the order in which `updateTreatments` is called would change the order of the keys in these objects. The "normal" es6 JS-Set also guarantees to return all elements in insertion order. Though in reality, this shouldn't matter.

TLDR: I'd also be fine with removing this comment if you think it won't be necessary for future programmers.

Reply:
The comment could stay there.
But the problem when this happens:
Right now the treatment calculation is independent from the history, resulting in when people drag something around it doesn't matter the order they drag personnel, material and patients around. Each time they will get the exact same.
Right now the assignment algorithm is stable and can therefore do incremental updates without any dependence on the history of updates and getting the same result.

If it is okay for people to have different assignments when the history is different, then it can stay. It is reasonable for this to happen, as personnel will probably not switch between two patients having the same category and stay with the one they already treat (maybe until the patient is stable or so).

But I would at least write in the comment that the algorithm is right now stable and that if things change so it is not anymore, it should be done not just carelessly but a be a design choice and implications should be considered.

E.g. an export could happen with all assigments deleted (as long as the spatialTrees are filled, which actually could also be rebuild at import with the mass import function of rbush) and recalculated on import and would result in the exact same assignements. Before export

Reply:
This was a design choice. I asked you whether this would be ok when I started reviewing this PR.

This comment should be removed if this isn't a design choice anymore (because now the algorithm is seen as by design stable and not just by chance).

Reply:
I would remove the part, but keep the part with not being perfectly realistic assignment (or how it is right now written)

Reply:
Continuation in https://github.com/hpi-sam/digital-fuesim-manv/pull/523/files/76eaac75c386627cb4ad0d9bb9f8da05540ccb92#r983319421